js/codemeta_generation.js

```diff
@@ -115,49 +128,65 @@ function generatePersons(prefix) {
+        "@type": "SoftwareSourceCode",
+    };
 
-        // Generate most fields
-        directCodemetaFields.forEach(function (item, index) {
-            doc[item] = getIfSet('#' + item)
-        });
+    let licenses = getLicenses();
+    if (licenses.length > 0) {
+        doc["license"] = (licenses.length === 1) ? licenses[0] : licenses;
+    }
 
-        doc["funder"] = generateShortOrg('#funder', doc["affiliation"]);
+    // Generate most fields
+    directCodemetaFields.forEach(function (item, index) {
+        doc[item] = getIfSet('#' + item)
+    });
 
-        // Generate simple fields parsed simply by splitting
-        splittedCodemetaFields.forEach(function (item, index) {
-            const id = item[0];
-            const separator = item[1];
-            const value = getIfSet('#' + id);
-            if (value !== undefined) {
-                doc[id] = value.split(separator).map(trimSpaces);
-            }
-        });
+    doc["funder"] = generateShortOrg('#funder', doc["affiliation"]);
 
-        // Generate dynamic fields
-        var authors = generatePersons('author');
-        if (authors.length > 0) {
-            doc["author"] = authors;
-        }
-        var contributors = generatePersons('contributor');
-        if (contributors.length > 0) {
-            doc["contributor"] = contributors;
+    // Generate simple fields parsed simply by splitting
+    splittedCodemetaFields.forEach(function (item, index) {
+        const id = item[0];
+        const separator = item[1];
+        const value = getIfSet('#' + id);
+        if (value !== undefined) {
+            doc[id] = value.split(separator).map(trimSpaces);
         }
+    });
 
-        codemetaText = JSON.stringify(doc, null, 4);
+    // Generate dynamic fields
+    var authors = generatePersons('author');
+    if (authors.length > 0) {
+        doc["author"] = authors;
+    }
+    var contributors = generatePersons('contributor');
+    if (contributors.length > 0) {
+        doc["contributor"] = contributors;
+    }
+    return doc;
+}
+
+// FIXME this hack to make expansion work for now
+// (expansion does not accept empty values)
+function deleteEmptyValues(doc) {
+  return JSON.parse(JSON.stringify(doc));
+}
+
+async function generateCodemeta() {
+    var inputForm = document.querySelector('#inputForm');
+    var codemetaText, codemetaTextToValidate, errorHTML;
+
+    if (inputForm.checkValidity()) {
+        var doc = buildDoc();
+        const expanded = await jsonld.expand(deleteEmptyValues(doc), {documentLoader: jsonldCustomLoader});
+        const compacted = await jsonld.compact(expanded, CODEMETA_CONTEXT_URL, {documentLoader: jsonldCustomLoader});
+        codemetaText = JSON.stringify(compacted, null, 4);
+         // FIXME the hack below to make validation work for now
+        //  (use manually built doc instead of compacted doc for validation)
+        codemetaTextToValidate = JSON.stringify(doc, null, 4);
```

Comment:
How can I reproduce this issue?

Reply:
You can replace with `codemetaTextToValidate = codemetaText;` to reproduce.
For example, the test `One full author can be exported` fail with the error `Unknown field "familyName" in "author".`. From my first searches, it comes from the fact that the expanded codemeta no longer has the same fields anymore (not the same amount, since I have deleted empty values, and not in the same order).

Reply:
Phew, that was a tough one. It was a bug in the existing code, and it was triggered here because the order of iteration changed. Fix here: https://github.com/codemeta/codemeta-generator/pull/28